File: spatial-lib/src/cmr/spatial/vector.clj
(ns cmr.spatial.vector
  (:require [cmr.spatial.math :refer :all]
            [primitive-math]
            [cmr.common.dev.record-pretty-printer :as record-pretty-printer]))
(primitive-math/use-primitive-operators)

(defrecord Vector
  [
   ^double x
   ^double y
   ^double z])

(record-pretty-printer/enable-record-pretty-printing Vector)

(defn new-vector
  [x y z]
  (->Vector (double x) (double y) (double z)))

(defn length ^double [^Vector v]
  (let [x (.x v) y (.y v) z (.z v)]
    (sqrt (+ (* x x) (* y y) (* z z)))))

(defn normalize [^Vector v]
  (let [l (length v)
        x (.x v) y (.y v) z (.z v)]
    (if (= 0.0 l)
      (throw (Exception. "Cannot normalize a vector length 0"))
      (new-vector (/ x l)
                  (/ y l)
                  (/ z l)))))

(defn cross-product [^Vector v1 ^Vector v2]
  (let [x1 (.x v1) y1 (.y v1) z1 (.z v1)
        x2 (.x v2) y2 (.y v2) z2 (.z v2)

        px (- (* y1 z2) (* y2 z1))
        py (- (* z1 x2) (* z2 x1))
        pz (- (* x1 y2) (* y1 x2))]
    (new-vector px py pz)))

(defn opposite
  "Finds the vector pointing in the opposite direction."
  [^Vector v]
  (let [x (.x v) y (.y v) z (.z v)]
    (new-vector (* -1.0 x) (* -1.0 y) (* -1.0 z))))

(def ^:const ^double VECTOR_EQUAL_DELTA
  "Defines how close parts of a vector can be considered before they will be considered equal. This
  is roughly equivalent to a 0.9 meters on the surface of the Earth."
  0.0000001)

(defn- vector-approx=
  ([^Vector v1 ^Vector v2]
   (vector-approx= v1 v2 VECTOR_EQUAL_DELTA))
  ([^Vector v1 ^Vector v2 ^double delta]
   (let [x1 (.x v1) y1 (.y v1) z1 (.z v1)
         x2 (.x v2) y2 (.y v2) z2 (.z v2)]
     (and
       (double-approx= x1 x2 delta)
       (double-approx= y1 y2 delta)
       (double-approx= z1 z2 delta)))))

(defn parallel?
  "Returns true if two vectors are parallel. Two vectors are parallel if they point in the same direction
  or opposite directions. Assumes both vectors are normalized"
  [v1 v2]
  (or (vector-approx= v1 v2)
      (vector-approx= (opposite v1) v2)))

File: spatial-lib/src/java/cmr/spatial/math/Vector.java
package cmr.spatial.math;

/**
 * Represents a 3D vector for spherical geometry calculations.
 * Immutable value class.
 */
public final class Vector {
    private final double x;
    private final double y;
    private final double z;

    public Vector(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public double getZ() {
        return z;
    }

    /**
     * Calculates the length (magnitude) of this vector.
     */
    public double length() {
        return Math.sqrt(x * x + y * y + z * z);
    }

    /**
     * Returns a normalized version of this vector (length = 1).
     * Throws IllegalArgumentException if vector length is 0.
     */
    public Vector normalize() {
        double len = length();
        if (len == 0.0) {
            throw new IllegalArgumentException("Cannot normalize a vector with length 0");
        }
        return new Vector(x / len, y / len, z / len);
    }

    /**
     * Calculates the cross product of this vector with another.
     * Returns a vector perpendicular to both input vectors.
     */
    public Vector crossProduct(Vector other) {
        double px = this.y * other.z - other.y * this.z;
        double py = this.z * other.x - other.z * this.x;
        double pz = this.x * other.y - this.y * other.x;
        return new Vector(px, py, pz);
    }

    /**
     * Returns a vector pointing in the opposite direction.
     */
    public Vector opposite() {
        return new Vector(-x, -y, -z);
    }

    /**
     * Vector equality tolerance: roughly equivalent to 0.9 meters on Earth's surface.
     */
    public static final double VECTOR_EQUAL_DELTA = 0.0000001;

    /**
     * Tests if two vectors are approximately equal within default tolerance.
     */
    public boolean approxEquals(Vector other) {
        return approxEquals(other, VECTOR_EQUAL_DELTA);
    }

    /**
     * Tests if two vectors are approximately equal within specified tolerance.
     */
    public boolean approxEquals(Vector other, double delta) {
        return MathUtils.doubleApprox(this.x, other.x, delta)
            && MathUtils.doubleApprox(this.y, other.y, delta)
            && MathUtils.doubleApprox(this.z, other.z, delta);
    }

    /**
     * Returns true if two vectors are parallel (point in same or opposite directions).
     * Assumes both vectors are normalized.
     */
    public boolean isParallel(Vector other) {
        return this.approxEquals(other) || this.opposite().approxEquals(other);
    }

    @Override
    public String toString() {
        return String.format("Vector{x=%.6f, y=%.6f, z=%.6f}", x, y, z);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Vector)) return false;
        Vector other = (Vector) obj;
        return Double.compare(x, other.x) == 0
            && Double.compare(y, other.y) == 0
            && Double.compare(z, other.z) == 0;
    }

    @Override
    public int hashCode() {
        int result = 17;
        long temp = Double.doubleToLongBits(x);
        result = 31 * result + (int)(temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(y);
        result = 31 * result + (int)(temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(z);
        result = 31 * result + (int)(temp ^ (temp >>> 32));
        return result;
    }
}
